### ISSUE TITLE
AdaLAM fails on identical images

### ISSUE THREAD

**Issue by vakker** [kornia/kornia#2221]
### Describe the bug

I'm running the the AdaLAM tutorial: https://kornia-tutorials.readthedocs.io/en/latest/image_matching_adalam.html
I'm doing some testing on rotating a reference image. When the 2 images are identical then AdaLAM provides an empty tensor for the output. 

### Reproduction steps

```bash
1. Check https://kornia-tutorials.readthedocs.io/en/latest/image_matching_adalam.html
2. Change 

fname1 = 'kn_church-2.jpg'
fname2 = 'kn_church-2.jpg'

3. `KF.match_adalam` will output:
```python
>>> print(dists)
tensor([], device='cuda:0', size=(0, 1))
>>> print(idxs)
tensor([], device='cuda:0', size=(0, 2), dtype=torch.int64)
```
```


### Expected behavior

Other approached (LoFTR, etc.) return the matches as expected. 
This is an issue, because often I have images that are only slightly rotated or translated, and the pipeline would fail if an image pair comes up that are very close. 
Also, it's unclear whether this is an issue with the algorithm itself, or some implementation bug.

### Environment

```shell
Collecting environment information...
PyTorch version: 1.13.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.20.5
Libc version: glibc-2.31

Python version: 3.9.7 (default, Nov  9 2021, 21:35:11)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-128-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.5.50
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce GTX 1080 Ti
GPU 1: NVIDIA GeForce GTX 1080 Ti

Nvidia driver version: 495.29.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.3.2
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   43 bits physical, 48 bits virtual
CPU(s):                          24
On-line CPU(s) list:             0-23
Thread(s) per core:              2
Core(s) per socket:              12
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           1
Model name:                      AMD Ryzen Threadripper 1920X 12-Core Processor
Stepping:                        1
CPU MHz:                         3597.595
BogoMIPS:                        6999.16
Virtualisation:                  AMD-V
L1d cache:                       384 KiB
L1i cache:                       768 KiB
L2 cache:                        6 MiB
L3 cache:                        32 MiB
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid amd_dcm aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate sme ssbd sev ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 xsaves clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.13.1
[conda] Could not collect
```
```


### Additional context

_No response_

**Comment by edgarriba** [comment]
This one is very interesting. Just for sanity, have you tried to apply a small transform to one of the samples ? /cc @ducha-aiki

**Comment by vakker** [comment]
Yes, I did. It's very weird. 
<details>
<summary>I used (same as the tutorial with minor changes)</summary>

```python
def get_matching_keypoints(lafs1, lafs2, idxs):
    mkpts1 = KF.get_laf_center(lafs1).squeeze()[idxs[:,0]].detach().cpu().numpy()
    mkpts2 = KF.get_laf_center(lafs2).squeeze()[idxs[:,1]].detach().cpu().numpy()
    return mkpts1, mkpts2

f_0 = "kn_church-2.jpg"
img_0 = load_torch_image(f_0)
img_0 = K.geometry.resize(img_0, 640, antialias=True)[:, :, :, :640]
img_1_orig = img_0

use_cuda: bool = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
print("Device:", device)
img_0_gray = K.color.rgb_to_grayscale(img_0).to(device)

feature = KF.KeyNetAffNetHardNet(5000, True).eval().to(device)

results = []
total_rot = 5
for i in range(total_rot):
    angle = torch.tensor([i * 5/total_rot])
    img_1 = transform.Rotate(angle)(img_1_orig)
    img_1_gray = K.color.rgb_to_grayscale(img_1).to(device)
    
    hw0 = torch.tensor(img_0.shape[2:])
    hw1 = torch.tensor(img_1.shape[2:])
    
    
    with torch.inference_mode():
        lafs0, resps0, descs0 = feature(img_0_gray)
        lafs1, resps1, descs1 = feature(img_1_gray)
        dists, idxs = KF.match_adalam(descs0.squeeze(0), 
                                      descs1.squeeze(0),
                                      lafs0,
                                      lafs1,
                                      hw1=hw0,
                                      hw2=hw1)
    #print(f"{idxs.shape[0]} tentative matches with AdaLAM")

    mkpts0, mkpts1 = get_matching_keypoints(lafs0, lafs1, idxs)
    
    if idxs.shape[0]>0:
        Fm, inliers = cv2.findFundamentalMat(
            mkpts0, mkpts1, cv2.USAC_MAGSAC, 0.75, 0.999, 100000
        )
        inliers = inliers > 0
    else:
        inliers = None
            

    
    result = {
        'img0': img_0,
        'img1': img_1,
        'lafs0': lafs0.cpu(),
        'lafs1': lafs1.cpu(),
        'tent_idxs': idxs.cpu(),
        'mkpts0': mkpts0,
        'mkpts1': mkpts1,
        'inliers':inliers,
        'model': model,
    }
    
    results.append(result)

def plot(idx):
    result = results[idx]
    mkpts0 = result['mkpts0']
    mkpts1 = result['mkpts1']
    lafs0 = result['lafs0']
    lafs1 = result['lafs1']
    img0 = result['img0']
    img1 = result['img1']
    tent_idxs = result['tent_idxs']
    inliers = result['inliers']
    
    if inliers is None:
        inliers = np.ones((len(mkpts0), 1))
    
    draw_LAF_matches(
        lafs0,
        lafs1,
        tent_idxs,
        K.tensor_to_image(img0),
        K.tensor_to_image(img1),
        inliers,
        draw_dict={
            "inlier_color": (0.2, 1, 0.2, 0.1),
            #'tentative_color': (1, 1, 0.2, 0.3),
            #"feature_color": (0.2, 0.5, 1, 0),
            "feature_color": None,
            'tentative_color': None,
            "vertical": False,
        },
    )
```
</details>

For `i=0`:
![test1](https://user-images.githubusercontent.com/3268882/220208762-b505c574-bc09-4863-a77f-d5c352bfc42f.png)

For `i=1`:
![test2](https://user-images.githubusercontent.com/3268882/220208796-ffb44c6e-12dc-4447-a07f-1dde4bd9febc.png)

**Comment by xmba15** [comment]
I encountered the same phenomenon when using adalam...

I'd like to also tag the original author here to see if he has any insights on this.
@cavalli1234

**Comment by johnnv1** [comment]
From #2226 
> Not yet now, though. The seeming reason for failing to match same image, is because https://github.com/kornia/kornia/blob/master/kornia/feature/adalam/ransac.py#L58 AdaLAM RANSAC filters out "too perfect fits" and then some division by zero happening as well.
It is not clear where exactly fix should be.

The test case showing the issue created on the PR:
```pytest
    @pytest.mark.parametrize("data", ["adalam_idxs"], indirect=True)
    def test_same_image(self, device, dtype, data):
        torch.random.manual_seed(0)
        # This is not unit test, but that is quite good integration test
        data_dev = utils.dict_to(data, device, dtype)
        matcher = GeometryAwareDescriptorMatcher('adalam', {"device": device}).to(device, dtype)
        with torch.no_grad():
            dists, idxs = matcher(data_dev['descs1'], data_dev['descs1'], data_dev['lafs1'], data_dev['lafs1'])
        assert idxs.shape[1] == 2
        assert dists.shape[1] == 1
        assert idxs.shape[0] == dists.shape[0]
        expected_idxs = torch.arange(data_dev['descs1'].shape[0]).to(device).view(-1, 1).repeat(1, 2)
        assert_close(idxs, expected_idxs, rtol=1e-4, atol=1e-4)
```
